edx_sga/sga.py

```diff
@@ -227,67 +232,55 @@ def student_state(self):
-                course_id=self.course_id,
-                item_id=self.block_id)
-            for student in students:
-                submission = self.get_submission(student.student_id)
-                user = user_by_anonymous_id(student.student_id)
-                if not submission or user.is_staff or user.is_superuser:
-                    continue
-                module, _ = StudentModule.objects.get_or_create(
-                    course_id=self.course_id,
-                    module_state_key=self.location,
-                    student=user,
-                    defaults={
-                        'state': '{}',
-                        'module_type': self.category,
-                    })
-                state = json.loads(module.state)
-                score = self.get_score(student.student_id)
-                downloaded = self.get_submission_download_status(student.student_id)
-                yield {
-                    'module_id': module.id,
-                    'student_id': student.student_id,
-                    'submission_id': submission['uuid'],
-                    'username': module.student.username,
-                    'fullname': module.student.profile.name,
-                    'filename': submission['answer']["filename"],
-                    'downloaded': downloaded,
-                    'timestamp': str(submission['created_at']),
-                    'score': score,
-                    'annotated': state.get("annotated_filename"),
-                    'comment': state.get("comment", ''),
-                }
-
-        enrolled_students = CourseEnrollment.objects.users_enrolled_in(self.course_id).exclude(
-            Q(is_staff=True) | Q(is_superuser=True)
-        )
-        submitted_student_data = list(get_student_data())
-        submitted_student_ids = list(map((lambda x: user_by_anonymous_id(x['student_id']).id), submitted_student_data))
-        not_submitted_students = enrolled_students.exclude(id__in=submitted_student_ids)
-
-        assignments = submitted_student_data
-        for student in not_submitted_students:
-            assignments.append({
-                'module_id': None,
-                'student_id': student.id,
-                'submission_id': None,
-                'username': student.username,
-                'fullname': student.profile.name,
-                'filename': None,
-                'downloaded': False,
-                'timestamp': None,
-                'score': None,
-                'annotated': None,
-                'comment': None,
-            })
+        def get_student_data(enrolled_students = None):
+            if not enrolled_students:
+                enrolled_students = []
+
+            submitted_student_data = []
+            for user in enrolled_students:
+                module = None
+                submission = None
+                user.student_id = anonymous_id_for_user(user, self.course_id, save=False)
```

Comment:
Once you remove `generate_student_data` you don't need to append attributes to object :mask: